src/producer/transactionManager.js

```diff
@@ -0,0 +1,97 @@
+const NO_PRODUCER_ID = -1
+const SEQUENCE_START = 0
+
+/**
+ * Manage behavior for an idempotent producer and transactions.
+ */
+module.exports = ({ logger, cluster, transactionTimeout = 30000 }) => {
+  /**
+   * Current producer ID
+   */
+  let producerId = NO_PRODUCER_ID
+
+  /**
+   * Current producer epoch
+   */
+  let producerEpoch = 0
+
+  /**
+   * Idempotent production requires that the producer track the sequence number of messages.
+   *
+   * Sequences are sent with every Record Batch and tracked per Topic-Partition
+   */
+  let producerSequence = {}
+
+  const isInitialized = () => producerId !== NO_PRODUCER_ID
+
+  return {
+    /**
+     * Initialize the idempotent producer by making an `InitProducerId` request.
+     * Overwrites any existing state in this transaction manager
+     */
+    initProducerId: async () => {
+      await cluster.refreshMetadataIfNecessary()
+
+      // If non-transactional we can request the PID from any broker
+      const broker = await cluster.findControllerBroker()
+      const result = await broker.initProducerId({ transactionTimeout })
+
+      producerId = result.producerId
```

Comment:
Shouldn't we preserve this across reconnections?

Reply:
@tulios Yes I think we should, good point! I'll update the producer so that it only calls `initProducerId` if uninitialized.

Reply:
@ianwsperber will you do this on the transactional producer PR or this one? (If you decide for the transactional PR we can merge)

Reply:
@tulios Forgot about this, just pushed a commit to address